### PR TITLE
Infer required Delta Lake version from table features

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

Rather than require users upgrade the protocol version on their tables before using features which require newer versions of the spec, automatically determine which protocol version is required and upgrade the table to it.

## Additional context and related issues

Includes one commit from @homar's https://github.com/trinodb/trino/pull/16221

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Infer required Delta Lake version when creating new tables or enabling new table features.
```

### CODE REVIEW

**Comment by homar** [comment]
Why didn't you add `testTurningOnAndOffCdfFromTrino` that I added to `testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDatabricksChangeDataFeedCompatibility.java` ?

**Comment by alexjo2144** [comment]
Because the commit message made it sound like I could skip it "Add flaky to all tests...", maybe it should be in the second commit, or a separate one all together?

**Comment by homar** [comment]
> Because the commit message made it sound like I could skip it "Add flaky to all tests...", maybe it should be in the second commit, or a separate one all together?

I added `Flaky` annotations to all methods in that class in a first commit and it made sense to skip that commit, but in the second commit that you actually used I added a test case there, and @ebyhr  thought it is important because he wanted me to test some particular cases there so I think that one test should be here.

I will add @Flaky to all tests in that class in a separate pr

**Review by homar — APPROVED** [review]

File: docs/src/main/sphinx/connector/delta-lake.rst [hunk 1]
```diff
@@ -573,10 +563,6 @@ The following properties are available for use:
     - Set the checkpoint interval in seconds.
   * - ``change_data_feed_enabled``
     - Enables storing change data feed entries.
-  * - ``reader_version``
-    - Set reader version.
-  * - ``writer_version``
-    - Set writer version.
 
 The following example uses all six table properties::
```
Comment (krvikash) [inline docs/src/main/sphinx/connector/delta-lake.rst:567]: `six` -> `four`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 2]
```diff
@@ -484,34 +476,4 @@ public DeltaLakeConfig setRegisterTableProcedureEnabled(boolean registerTablePro
         this.registerTableProcedureEnabled = registerTableProcedureEnabled;
         return this;
     }
-
-    @Min(value = MIN_READER_VERSION, message = "Must be in between " + MIN_READER_VERSION + " and " + MAX_READER_VERSION)
-    @Max(value = MAX_READER_VERSION, message = "Must be in between " + MIN_READER_VERSION + " and " + MAX_READER_VERSION)
-    public int getDefaultReaderVersion()
```
Comment (krvikash) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:490]: [TestDeltaLakePageSink.java](https://github.com/trinodb/trino/pull/16165/files#diff-66264364c4b34e57284c11079858ad05f15bf7e5ad63b06f3d3e18646efaa4da) and [TestDeltaLakeConfig.java](https://github.com/trinodb/trino/pull/16208/files#diff-bba17fc0643cf731516d7bf56e5736dd1e6c8dad8aeb62a52fa1758d3acdb4d3) needs to be updated.

Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:490]: [DeltaLakeMetadataFactory.java](https://github.com/trinodb/trino/pull/16208/files#diff-d50f059e35f4a166854e3c70c2a5ba77f3abaa9dbd1e7caad4aef1371b454e60) needs to be updated.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 3]
```diff
@@ -287,6 +284,11 @@
     public static final String CHANGE_COLUMN_OPERATION = "CHANGE COLUMN";
     public static final String ISOLATION_LEVEL = "WriteSerializable";
 
+    public static final int DEFAULT_READER_VERSION = 1;
+    public static final int DEFAULT_WRITER_VERSION = 2;
+    // The highest reader and writer versions Trino supports writing to
+    public static final int MAX_READER_VERSION = 2;
+    public static final int MAX_WRITER_VERSION = 4;
     private static final int CDF_SUPPORTED_WRITER_VERSION = 4;
```
Comment (krvikash) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:292]: `defaultReaderVersion` and `defaultWriterVersion` needs to be removed from `DeltaLakeMetadata`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -287,6 +284,11 @@
     public static final String CHANGE_COLUMN_OPERATION = "CHANGE COLUMN";
     public static final String ISOLATION_LEVEL = "WriteSerializable";
 
+    public static final int DEFAULT_READER_VERSION = 1;
+    public static final int DEFAULT_WRITER_VERSION = 2;
+    // The highest reader and writer versions Trino supports writing to
+    public static final int MAX_READER_VERSION = 2;
```
Comment (krvikash) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:290]: unused `MAX_READER_VERSION`

File: docs/src/main/sphinx/connector/delta-lake.rst (hunk 1 above)
Comment (findepi) [reply docs/src/main/sphinx/connector/delta-lake.rst:567]: good catch. very tricky

"The following example uses multiple table properties" would avoid having to update that line.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 5]
```diff
@@ -1908,20 +1909,44 @@ public void setTableProperties(ConnectorSession session, ConnectorTableHandle ta
                         "%s cannot be downgraded from %d to %d", WRITER_VERSION_PROPERTY, currentProtocolEntry.getMinWriterVersion(), writerVersion.get()));
             }
         }
+        long createdTime = Instant.now().toEpochMilli();
+
+        Optional<MetadataEntry> metadataEntry = Optional.empty();
+        if (properties.containsKey(CHANGE_DATA_FEED_ENABLED_PROPERTY)) {
+            boolean changeDataFeedEnabled = (Boolean) properties.get(CHANGE_DATA_FEED_ENABLED_PROPERTY)
+                    .orElseThrow(() -> new IllegalArgumentException("The change_data_feed_enabled  property cannot be empty"));
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1883]: double space

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 6]
```diff
@@ -1908,20 +1909,44 @@ public void setTableProperties(ConnectorSession session, ConnectorTableHandle ta
...
+            long latestWriterVersion = writerVersion.orElse(currentProtocolEntry.getMinWriterVersion());
+            if (changeDataFeedEnabled && (latestWriterVersion < CDF_SUPPORTED_WRITER_VERSION)) {
+                throw new TrinoException(
+                        INVALID_TABLE_PROPERTY,
+                        format(
+                                "%s is set to less than %s which is a minimal %s that supports %s",
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java]: "%s is set to %s and %s is required to enable change data feed"

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 7]
```diff
@@ -1930,6 +1955,19 @@ public void setTableProperties(ConnectorSession session, ConnectorTableHandle ta
         }
     }
 
+    private MetadataEntry buildMetadataEntry(MetadataEntry metadataEntry, Map<String, String> configuration, long createdTime)
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1914]: is `createdTime` time of table creation or time of metadata entry creation?

(in the latter case, the code here could take current time itself)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 8]
```diff
@@ -1930,6 +1955,19 @@ public void setTableProperties(ConnectorSession session, ConnectorTableHandle ta
...
+    {
+        return new MetadataEntry(
+                metadataEntry.getId(),
+                metadataEntry.getName(),
+                metadataEntry.getDescription(),
+                metadataEntry.getFormat(),
+                metadataEntry.getSchemaString(),
+                metadataEntry.getOriginalPartitionColumns(),
+                ImmutableMap.copyOf(configuration),
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1923]: no need to copyOf here. it's MetadataEntry constructor's responsibility to do any defensive copies it needs

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 7 above)
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1914]: It is set in both the `MetadataEntry` and `CommitInfoEntry`, we're doing this so the two have the same time.

**Comment by alexjo2144** [comment]
Thanks for the reviews, all addressed

**Review by krvikash — APPROVED** [review]
LGTM

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 9]
```diff
@@ -1889,39 +1867,35 @@ public void setTableProperties(ConnectorSession session, ConnectorTableHandle ta
...
-            writerVersion = Optional.of((int) properties.get(WRITER_VERSION_PROPERTY)
-                    .orElseThrow(() -> new IllegalArgumentException("The " + WRITER_VERSION_PROPERTY + " property cannot be empty")));
-            if (writerVersion.get() < currentProtocolEntry.getMinWriterVersion()) {
-                throw new TrinoException(INVALID_TABLE_PROPERTY, format(
-                        "%s cannot be downgraded from %d to %d", WRITER_VERSION_PROPERTY, currentProtocolEntry.getMinWriterVersion(), writerVersion.get()));
+        int requiredWriterVersion = currentProtocolEntry.getMinWriterVersion();
+        Optional<MetadataEntry> metadataEntry = Optional.empty();
+        if (properties.containsKey(CHANGE_DATA_FEED_ENABLED_PROPERTY)) {
+            boolean changeDataFeedEnabled = (Boolean) properties.get(CHANGE_DATA_FEED_ENABLED_PROPERTY)
+                    .orElseThrow(() -> new IllegalArgumentException("The change_data_feed_enabled  property cannot be empty"));
```
Comment (krvikash) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1876]: double space

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
cc @trinodb/maintainers @colebow  i marked this one as a release-blocker, because it's important to get this into the release.
This is not a typical release blocker. It removes some recently introduced properties (https://github.com/trinodb/trino/pull/16332) and the goal here is to minimize the time window during which these are available to effectively minimize any negative consequences.

**Pull request merged**